src/PostgREST/Request/ApiRequest.hs

```diff
@@ -448,10 +469,46 @@ requestContentTypes conf action target =
     invokeContentTypes =
       defaultContentTypes
         ++ rawContentTypes conf
-        ++ [CTOpenAPI | tpIsRootSpec target]
+        ++ [CTOpenAPI | pIsRootSpec path]
     defaultContentTypes =
       [CTApplicationJSON, CTSingularJSON, CTTextCSV]
 
 rawContentTypes :: AppConfig -> [ContentType]
 rawContentTypes AppConfig{..} =
   (ContentType.decodeContentType <$> configRawMediaTypes) `union` [CTOctetStream, CTTextPlain]
+
+{-|
+  Search a pg procedure by its parameters. Since a function can be overloaded, the name is not enough to find it.
+  An overloaded function can have a different volatility or even a different return type.
+-}
+findProc :: QualifiedIdentifier -> S.Set Text -> Bool -> ProcsMap -> Either ApiRequestError ProcDescription
+findProc qi payloadKeys paramsAsSingleObject allProcs =
+  case bestMatch of
+    []     -> Left $ NoRpc (qiSchema qi) (qiName qi) (S.toList payloadKeys) paramsAsSingleObject
+    [proc] -> Right proc
+    procs  -> Left $ AmbiguousRpc (toList procs)
+  where
+    bestMatch =
+      case M.lookup qi allProcs of
+        Nothing     -> []
+        Just [proc] -> [proc | matches proc]
+        Just procs  -> filter matches procs
+    -- Find the exact arguments match
+    matches proc
+      | paramsAsSingleObject = case pdArgs proc of
+                               [arg] -> pgaType arg `elem` ["json", "jsonb"]
+                               _     -> False
+      | otherwise            = case pdArgs proc of
+                               []   -> null payloadKeys
+                               args -> matchesArg args
+    matchesArg args =
+      -- The function's required arguments are separated from the ones with a default value assigned.
+      -- The set of names of those arguments is compared to the set of keys supplied by the client
+      -- 1. If only required arguments are found, the keys must be exactly the same as those arguments
+      -- 2. If only optional arguments are found, the keys must be a subset of those arguments
+      -- 3. If both required and optional arguments are found, the result of taking away the optional arguments
+      --    from the keys must be exactly the same as the required arguments
+      case L.partition pgaReq args of
+        (reqArgs, [])      -> payloadKeys == S.fromList (pgaName <$> reqArgs)
+        ([], defArgs)      -> payloadKeys `S.isSubsetOf` S.fromList (pgaName <$> defArgs)
+        (reqArgs, defArgs) -> payloadKeys `S.difference` S.fromList (pgaName <$> defArgs) == S.fromList (pgaName <$> reqArgs)
```

Comment:
@laurenceisla This is the only part that I still need to check. I'll run some bench tests to see if there's an impact on RPC performance. Ideally, the non-overloaded case shouldn't suffer a perf penalty.

Reply:
I've run the benchmark tests and these are the results. 

```
pgrbench-k6 t3anano k6/RPCSimple.js
    http_reqs..................: 55521  1844.785613/s
    http_reqs..................: 56208  1868.872353/s
    http_reqs..................: 56631  1871.617673/s

pgrbench-k6 t3anano k6/RPCGETSingle.js
    http_reqs..................: 61239  2030.219531/s
    http_reqs..................: 60722  2001.174361/s
    http_reqs..................: 60647  2016.992396/s

pgrbench-k6 t3anano k6/RPCGETSingleEmbed.js
    http_reqs..................: 24001  798.671285/s
    http_reqs..................: 24001  794.321514/s
    http_reqs..................: 23997  796.039754/s
```

Compared to the default ones in the benchmark there's a small improvement in the RPCSimple test while the others give similar results. I'll build the latest nightly and see how they compare.

Reply:
These are the benchmarks for the latest nightly f3a184af017381129245db6c98743963af9b6d45 . I still need to resolve the merge conflicts with this branch and, if the results are too different, I'll update the benchmark I made here: https://github.com/PostgREST/postgrest/pull/1841#discussion_r661919724

```
pgrbench-k6 t3anano k6/RPCSimple.js
    http_reqs..................: 55615  1848.913494/s
    http_reqs..................: 55428  1842.921351/s
    http_reqs..................: 55743  1842.730914/s


pgrbench-k6 t3anano k6/RPCGETSingle.js
    http_reqs..................: 57051  1896.411094/s
    http_reqs..................: 58782  1940.730047/s
    http_reqs..................: 59174  1967.293803/s

pgrbench-k6 t3anano k6/RPCGETSingleEmbed.js
    http_reqs..................: 24001  798.712055/s
    http_reqs..................: 23979  797.954143/s
    http_reqs..................: 24000  798.706433/s
```

Reply:
@laurenceisla So according to your results [here](https://github.com/PostgREST/postgrest/pull/1841#discussion_r661919724), we're actually improving RPC perf!

Why is that? The above logic looked a bit more complicated.

Reply:
> These are the benchmarks for the latest nightly f3a184a . I still need to resolve the merge conflicts with this branch and, if the results are too different, I'll update the benchmark I made here: #1841 (comment)

@steve-chavez  I resolved the merge conflicts and redid the benchmarks. The results are now very similar to the nightly I mentioned here. So, it looks like the performance isn't better or worse.

> @laurenceisla So according to your results here, we're actually improving RPC perf!
>
> Why is that? The above logic looked a bit more complicated.

I thought it was because the `matchesArgs` function now verifies using equality for RPCs without optional parameters, instead of using the `isSubsetOf` function all the time. But with this last merge, it seems like it didn't affect the performance in any way.

Reply:
Ok, cool. Good to know it doesn't affect performance. I'll merge now.